sycl/include/sycl/ext/oneapi/experimental/user_defined_reductions.hpp

```diff
@@ -0,0 +1,148 @@
+                     sycl::detail::is_native_op<T, BinaryOperation>::value),
+                    T>
+reduce_over_group(GroupHelper g, T x, BinaryOperation binary_op) {
+  return sycl::reduce_over_group(g.get_group(), x, binary_op);
+}
+
+template <typename GroupHelper, typename V, typename T, class BinaryOperation>
+sycl::detail::enable_if_t<
+    (is_group_helper_v<std::decay_t<GroupHelper>> &&
+     (sycl::detail::is_scalar_arithmetic<V>::value ||
+      sycl::detail::is_complex<V>::value) &&
+     (sycl::detail::is_scalar_arithmetic<T>::value ||
+      sycl::detail::is_complex<T>::value) &&
+     sycl::detail::is_native_op<V, BinaryOperation>::value &&
+     sycl::detail::is_native_op<T, BinaryOperation>::value &&
+     sycl::detail::is_plus_if_complex<T, BinaryOperation>::value &&
+     sycl::detail::is_plus_if_complex<V, BinaryOperation>::value),
+    T>
+reduce_over_group(GroupHelper g, V x, T init, BinaryOperation binary_op) {
+  return sycl::reduce_over_group(g.get_group(), x, binary_op);
+}
+
+template <typename GroupHelper, typename V, typename T, class BinaryOperation>
+sycl::detail::enable_if_t<
+    (is_group_v<std::decay_t<GroupHelper>> &&
+     sycl::detail::is_vector_arithmetic<V>::value &&
+     sycl::detail::is_vector_arithmetic<T>::value &&
+     sycl::detail::is_native_op<V, BinaryOperation>::value &&
+     sycl::detail::is_native_op<T, BinaryOperation>::value),
+    T>
+reduce_over_group(GroupHelper g, V x, T init, BinaryOperation binary_op) {
+  return sycl::reduce_over_group(g.get_group(), x, init, binary_op);
+}
+
+// ---- joint_reduce
+template <typename GroupHelper, typename Ptr, typename BinaryOperation>
+sycl::detail::enable_if_t<(is_group_helper_v<std::decay_t<GroupHelper>> &&
+                           sycl::detail::is_pointer<Ptr>::value),
+                          typename sycl::detail::remove_pointer<Ptr>::type>
+joint_reduce(GroupHelper g, Ptr first, Ptr last, BinaryOperation binary_op) {
+#ifdef __SYCL_DEVICE_ONLY__
+  using T = typename sycl::detail::remove_pointer<Ptr>::type;
+  T init = 0;
+  return joint_reduce(g, first, last, init, binary_op);
+#else
+  (void)g;
+  (void)first;
+  (void)last;
+  (void)binary_op;
+  throw runtime_error("Group algorithms are not supported on host device.",
+                      PI_ERROR_INVALID_DEVICE);
+#endif
+}
+
+template <typename GroupHelper, typename Ptr, typename T,
+          typename BinaryOperation>
+sycl::detail::enable_if_t<(is_group_helper_v<std::decay_t<GroupHelper>> &&
+                           sycl::detail::is_pointer<Ptr>::value),
+                          T>
+joint_reduce(GroupHelper g, Ptr first, Ptr last, T init,
+             BinaryOperation binary_op) {
+#ifdef __SYCL_DEVICE_ONLY__
+  T partial;
+  sycl::detail::for_each(
+      g.get_group(), first, last,
+      [&](const typename sycl::detail::remove_pointer<Ptr>::type &x) {
+        partial = binary_op(partial, x);
+      });
```

Comment:
This seems to be linear and not logarithmic in complexity. I'd expect something like https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/reduction.hpp#L1810-L1818 might be applicable here, although I'm not sure what gives better performance for different work sizes.

Reply:
I would suggest to put a comment here for now. We don't yet have any data from actual uses of this new function about which use cases (data type, wg sizes) are the most popular ones.